### PR TITLE
[fix](function) make STRLEFT and STRRIGHT and SUBSTR function DEPEND_ON_ARGUMENT

### LINKED ISSUE #27435: [Enhancement] Reduct the unnecessary ALWAYS_NULLABLE in nullable mode
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

for many doris' functions, they have nullable mode `ALWAYS_NULLABLE` but in fact no need to be. many of them could be changed to `DEPEND_ON_ARGUMENT` or `CUSTOM` just like https://github.com/apache/doris/pull/27343, https://github.com/apache/doris/pull/22321 and https://github.com/apache/doris/pull/26827. 

### Solution

It may relate to 3 position: BE's function's implementation, function prototype of Nereids planner and function prototype of Legacy planner.
If could reduct it, reduct it.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by koarz**
## Proposed changes

Issue Number: #27435 

make STRLEFT and STRRIGHT and SUBSTR function DEPEND_ON_ARGUMENT

and the case where the strleft have NULL argument the result will always null(before get empty)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Review by zclllyybb — CHANGES_REQUESTED** [review]
and we need more regression-test

**Comment by koarz** [comment]
run buildall

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -581,7 +588,22 @@ class FunctionLeft : public IFunction {
         temp_arguments[0] = arguments[0];
         temp_arguments[1] = num_columns_without_result;
         temp_arguments[2] = arguments[1];
-        SubstringUtil::substring_execute(block, temp_arguments, result, input_rows_count);
+
+        bool is_nullable {};
```
Comment (zclllyybb) [inline be/src/vec/functions/function_string.h:592]: it will be more clear if change to
```suggestion
        bool is_nullable = nullptr;
```

**Comment by zclllyybb** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by zclllyybb — APPROVED** [review]
LGTM, thanks for your contribution!

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Review by zclllyybb — APPROVED** [review]

File: regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy [hunk 2]
```diff
@@ -140,8 +140,76 @@ suite("test_string_function", "arrow_flight_sql") {
     qt_sql "select starts_with(\"hello world\",\"world\");"
     qt_sql "select starts_with(\"hello world\",null);"
 
-    qt_sql "select strleft(\"Hello doris\",5);"
-    qt_sql "select strright(\"Hello doris\",5);"
+    qt_sql "select strleft(NULL, 1);"
+    qt_sql "select strleft(\"good morning\", NULL);"
+    qt_sql "select left(NULL, 1);"
```
Comment (BiteTheDDDDt) [inline regression-test/suites/query_p0/sql_functions/string_functions/test_string_function.groovy:145]: add case for invalid input

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by HappenLee** [comment]
should add version to get different function and keep old function to make update no cause error.

**Comment by koarz** [comment]
run pipelinep0

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run p0

**Comment by koarz** [comment]
run feut

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run p0

**Comment by koarz** [comment]
run pipeline_xp0

**Comment by koarz** [comment]
run pipelinex_p0

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run p0

**Comment by koarz** [comment]
run pipelinex_p0

**Review by zclllyybb — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**